extensions/amp-viewer-integration/0.1/amp-viewer-integration.js

```diff
@@ -47,53 +46,37 @@ export class AmpViewerIntegration {
   init() {
     dev().info(TAG, 'handshake init()');
     const viewer = viewerForDoc(this.win.document);
-    this.unconfirmedViewerOrigin_ = viewer.getParam('viewerorigin') || null;
+    this.unconfirmedViewerOrigin_ = viewer.getParam('viewerorigin');
     if (!this.unconfirmedViewerOrigin_) {
       dev().info(TAG, 'Viewer origin not specified.');
       return null;
     }
-    return this.getHandshakePromise_()
-      .then(viewerOrigin => {
-        dev().info(TAG, 'listening for messages');
-        const messaging =
-          new Messaging(this.win, this.win.parent, viewerOrigin,
-            (type, payload, awaitResponse) => {
-              return viewer.receiveMessage(
-                type, /** @type {!JSONType} */ (payload), awaitResponse);
-            });
-        viewer.setMessageDeliverer(messaging.sendRequest.bind(messaging),
-          viewerOrigin);
-      });
+
+    dev().info(TAG, 'listening for messages', this.unconfirmedViewerOrigin_);
+    const messaging = new Messaging(
+      this.win, this.win.parent, this.unconfirmedViewerOrigin_);
+
+    dev().info(TAG, 'Send a handshake request');
+    return this.openChannel(messaging)
+        .then(() => {
+          messaging.setRequestProcessor((type, payload, awaitResponse) => {
+            return viewer.receiveMessage(
+              type, /** @type {!JSONType} */ (payload), awaitResponse);
+          });
+
+          dev().info(TAG, 'setMessageDeliverer');
```

Comment:
Let's move to the top of `then` and say something like "channel has been opened".